farmerbot/server/farmerbot.go

```diff
@@ -0,0 +1,296 @@
+	network      string
+	mnemonic     string
+	identity     substrate.Identity
+}
+
+// NewFarmerBot generates a new farmer bot
+func NewFarmerBot(ctx context.Context, configFile, network, mnemonic string) (FarmerBot, error) {
+	farmerbot := FarmerBot{mnemonic: mnemonic, network: network, configFile: configFile}
+
+	substrateManager := substrate.NewManager(constants.SubstrateURLs[network]...)
+	sub, err := substrateManager.Substrate()
+	if err != nil {
+		return FarmerBot{}, err
+	}
+
+	farmerbot.sub = *sub
+
+	err = farmerbot.setConfig()
+	if err != nil {
+		return FarmerBot{}, err
+	}
+
+	// TODO: check farm hex seed if available
+	identity, err := substrate.NewIdentityFromSr25519Phrase(mnemonic)
+	if err != nil {
+		return FarmerBot{}, err
+	}
+
+	farmerbot.identity = identity
+
+	twinID, err := sub.GetTwinByPubKey(identity.PublicKey())
+	if err != nil {
+		return FarmerBot{}, err
+	}
+
+	farmerbot.twinID = twinID
+
+	powerManager := manager.NewPowerManager(identity, sub, farmerbot.config)
+	if err != nil {
+		return FarmerBot{}, err
+	}
+
+	farmerbot.powerManager = &powerManager
+
+	rmb, err := peer.NewRpcClient(ctx, peer.KeyTypeSr25519, mnemonic, constants.RelayURLS[network], fmt.Sprintf("farmerbot-rpc-%d", farmerbot.config.Farm.ID), sub, true)
+	if err != nil {
+		return FarmerBot{}, errors.Wrap(err, "could not create rmb client")
+	}
+
+	rmbNodeClient := manager.NewRmbNodeClient(rmb)
+	dataManager := manager.NewDataManager(sub, farmerbot.config, &rmbNodeClient)
+	if err != nil {
+		return FarmerBot{}, err
+	}
+
+	farmerbot.dataManager = dataManager
+
+	return farmerbot, nil
+}
+
+// Run runs farmerbot to update nodes and power management
+func (f *FarmerBot) Run(ctx context.Context) {
+	if err := f.start(ctx); err != nil {
+		log.Fatal().Err(err).Msg("[FARMERBOT] error starting")
+	}
+
+	go f.update(ctx)
+	go f.serve(ctx)
+
+	log.Info().Msg("[FARMERBOT] up and running...")
+
+	// graceful shutdown
+	sigChan := make(chan os.Signal, 1)
+	signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM)
+	<-sigChan
+
+	shutdownCtx, shutdownRelease := context.WithTimeout(context.Background(), 10*time.Second)
+	defer shutdownRelease()
+
+	if err := f.stop(shutdownCtx); err != nil {
+		log.Fatal().Err(err).Msg("[FARMERBOT] shutdown error")
+	}
+	log.Info().Msg("[FARMERBOT] graceful shutdown successful")
+}
+
+func (f *FarmerBot) setConfig() error {
+	content, format, err := parser.ReadFile(f.configFile)
+	if err != nil {
+		return err
+	}
+
+	inputs, err := parser.ParseIntoInputConfig(content, format)
+	if err != nil {
+		return err
+	}
+
+	if f.config == nil {
+		f.config = &models.Config{}
+	}
+
+	err = f.config.Set(&f.sub, inputs)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (f *FarmerBot) start(ctx context.Context) error {
+	return f.powerManager.PowerOnAllNodes()
+}
+
+func (f *FarmerBot) stop(ctx context.Context) error {
+	return f.powerManager.PowerOnAllNodes()
+}
+
+func (f *FarmerBot) update(ctx context.Context) {
```

Comment:
I can't explicelty state everything wrong in this approach but it's a general feeling of un-necessary compexity. The only concrete statemeng i can give is that:

Instead of doing everything in hidden subcalls that iterate over the entire shared state (this makes it really really hard to follow logic or debug) we should instead process each node in a single iteration

```
get list of ndoes from that farm that matches our config.
usedCap = 0
totalCap = 0
for node in nodes {
  - do we have a node state associdated with this node id. otherwise, fetch!
  - if it's off, do we need to wake it up for random wake ups? 
  - if node is on incrememnt both usedCap and totalCap
}
check percentage of usedCap to totalCap and then take action accordingly
```

this way the logic is straight forward and can be followed and debugged easily. 

> **NOTE** the problem as i see it is that you trying to follow the same design as the old one (with separate manager and so) which might not be necessary.

Reply:
Another thing that i find difficult about this function is you can't undesrand what it does by reading the function name, and logic.
for example what does f.dataManager.Update() means? or f.powerManager.PowerManager() what does that do. 

Hence the logic imho need to be cleaned and written as above, since it makes it easier to follow. It's not nice that u do the datamanager.Update() then do powermanager.Whatever() which has access to the shared state of the data manger.

I think splitting this into different manager make it harder and not improving abstration.